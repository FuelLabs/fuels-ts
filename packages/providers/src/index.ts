/// <reference types="graphql" />

export * from './coin-quantity';
export * from './coin';
export * from './provider';
export * from './message';
export * from './resource';
export { default as Provider } from './provider';
export * from './transaction-request';
export * from './transaction-response';
export * from './utils';
