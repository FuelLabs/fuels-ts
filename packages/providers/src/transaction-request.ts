import type { BigNumberish } from '@ethersproject/bignumber';
import { BigNumber } from '@ethersproject/bignumber';
import type { BytesLike } from '@ethersproject/bytes';
import { arrayify, hexlify } from '@ethersproject/bytes';
import type { Input, Output, Transaction, Witness } from '@fuel-ts/transactions';
import { TransactionType } from '@fuel-ts/transactions';

type ScriptTransactionRequest = {
  type: TransactionType.Script;
  gasPrice: BigNumberish;
  gasLimit: BigNumberish;
  maturity: BigNumberish;
  script: BytesLike;
  scriptData: BytesLike;
  inputs: Input[];
  outputs: Output[];
  witnesses: Witness[];
};

type CreateTransactionRequest = {
  type: TransactionType.Create;
  gasPrice: BigNumberish;
  gasLimit: BigNumberish;
  maturity: BigNumberish;
  bytecodeWitnessIndex: BigNumberish;
  salt: string;
  staticContracts: string[];
  inputs: Input[];
  outputs: Output[];
  witnesses: Witness[];
};

export type TransactionRequest = ScriptTransactionRequest | CreateTransactionRequest;

export const transactionFromRequest = (transactionRequest: TransactionRequest): Transaction => {
  switch (transactionRequest.type) {
    case TransactionType.Script: {
      const script = arrayify(transactionRequest.script);
      const scriptData = arrayify(transactionRequest.scriptData);
      const receiptsRoot = '0x0000000000000000000000000000000000000000000000000000000000000000';
      const inputs = transactionRequest.inputs;
      const outputs = transactionRequest.outputs;
      const witnesses = transactionRequest.witnesses;

      return {
        type: TransactionType.Script,
        data: {
          gasPrice: BigNumber.from(transactionRequest.gasPrice),
          gasLimit: BigNumber.from(transactionRequest.gasLimit),
          maturity: BigNumber.from(transactionRequest.maturity),
          scriptLength: BigNumber.from(script.length),
          scriptDataLength: BigNumber.from(scriptData.length),
          inputsCount: BigNumber.from(inputs.length),
          outputsCount: BigNumber.from(outputs.length),
          witnessesCount: BigNumber.from(witnesses.length),
          receiptsRoot,
          script: hexlify(script),
          scriptData: hexlify(scriptData),
          inputs,
          outputs,
          witnesses,
        },
      };
    }
    case TransactionType.Create: {
      const staticContracts = transactionRequest.staticContracts;
      const inputs = transactionRequest.inputs;
      const outputs = transactionRequest.outputs;
      const witnesses = transactionRequest.witnesses;
      const bytecodeWitnessIndex = BigNumber.from(transactionRequest.bytecodeWitnessIndex);
      const bytecodeLength = BigNumber.from(
        witnesses[bytecodeWitnessIndex.toNumber()].data.length / 4
      );

      return {
        type: TransactionType.Create,
        data: {
          gasPrice: BigNumber.from(transactionRequest.gasPrice),
          gasLimit: BigNumber.from(transactionRequest.gasLimit),
          maturity: BigNumber.from(transactionRequest.maturity),
          bytecodeLength,
          bytecodeWitnessIndex,
          staticContractsCount: BigNumber.from(staticContracts.length),
          inputsCount: BigNumber.from(inputs.length),
          outputsCount: BigNumber.from(outputs.length),
          witnessesCount: BigNumber.from(witnesses.length),
          salt: transactionRequest.salt,
          staticContracts,
          inputs,
          outputs,
          witnesses,
        },
      };
    }
    default: {
      throw new Error('Not implemented');
    }
  }
};
