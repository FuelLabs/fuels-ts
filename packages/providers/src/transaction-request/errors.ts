/* eslint-disable max-classes-per-file */
import type { AbstractAddress } from '@fuel-ts/interfaces';

export class ChangeOutputCollisionError extends Error {
  name = 'ChangeOutputCollisionError';
  message = 'A ChangeOutput with the same "assetId" already exists for a different "to" address';
}

export class NoWitnessAtIndexError extends Error {
  name = 'NoWitnessAtIndexError';
  constructor(public readonly index: number) {
    super();
    this.message = `Witness at index "${index}" was not found`;
  }
}

export class NoWitnessByOwnerError extends Error {
  name = 'NoWitnessByOwnerError';
  constructor(public readonly owner: AbstractAddress) {
    super();
    this.message = `A witness for the given owner "${owner}" was not found`;
  }
}
