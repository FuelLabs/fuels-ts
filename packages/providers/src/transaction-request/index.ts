export * from './input';
export * from './output';
export * from './transaction-request';
export * from './create-transaction-request';
export * from './script-transaction-request';
export * from './errors';
export * from './scripts';
export * from './types';
export * from './utils';
