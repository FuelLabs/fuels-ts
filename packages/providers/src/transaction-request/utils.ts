import { TransactionType } from '@fuel-ts/transactions';

import { CreateTransactionRequest } from './create-transaction-request';
import { ScriptTransactionRequest } from './script-transaction-request';
import type { TransactionRequestLike, TransactionRequest } from './types';

export const transactionRequestify = (obj: TransactionRequestLike): TransactionRequest => {
  if (obj instanceof ScriptTransactionRequest || obj instanceof CreateTransactionRequest) {
    return obj;
  }
  switch (obj.type) {
    case TransactionType.Script: {
      return ScriptTransactionRequest.from(obj);
    }
    case TransactionType.Create: {
      return CreateTransactionRequest.from(obj);
    }
    default: {
      throw new Error(
        `Unknown transaction type: ${
          // @ts-expect-error Unreachable code
          obj.type
        }`
      );
    }
  }
};
