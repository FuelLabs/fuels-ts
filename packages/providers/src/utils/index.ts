export * from './receipts';
export * from './fee';
export * from './block-explorer';
export * from './json';
export * from './sleep';
export * from './time';
