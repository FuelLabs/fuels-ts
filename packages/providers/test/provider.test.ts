import type { BytesLike } from '@ethersproject/bytes';
import { hexlify, arrayify } from '@ethersproject/bytes';
import { Address } from '@fuel-ts/address';
import { BaseAssetId, ZeroBytes32 } from '@fuel-ts/address/configs';
import { randomBytes } from '@fuel-ts/crypto';
import { safeExec } from '@fuel-ts/errors/test-utils';
import { BN, bn } from '@fuel-ts/math';
import type { Receipt } from '@fuel-ts/transactions';
import { InputType, ReceiptType, TransactionType } from '@fuel-ts/transactions';
import * as GraphQL from 'graphql-request';

import Provider from '../src/provider';
import type {
  CoinTransactionRequestInput,
  MessageTransactionRequestInput,
} from '../src/transaction-request';
import { ScriptTransactionRequest } from '../src/transaction-request';
import { fromTai64ToUnix, fromUnixToTai64 } from '../src/utils';

import { messageProofResponse } from './fixtures';

afterEach(() => {
  jest.restoreAllMocks();
});

// TODO: Figure out a way to import this constant from `@fuel-ts/wallet/configs`
const FUEL_NETWORK_URL = 'http://127.0.0.1:4000/graphql';

describe('Provider', () => {
  it('can getVersion()', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);

    const version = await provider.getVersion();

    expect(version).toEqual('0.20.4');
  });

  it('can call()', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);

    const CoinInputs: CoinTransactionRequestInput[] = [
      {
        type: InputType.Coin,
        id: '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c500',
        owner: BaseAssetId,
        assetId: BaseAssetId,
        txPointer: '0x00000000000000000000000000000000',
        amount: 100,
        witnessIndex: 0,
      },
    ];

    const callResult = await provider.call({
      type: TransactionType.Script,
      gasPrice: 0,
      gasLimit: 1000000,
      script:
        /*
          Opcode::ADDI(0x10, REG_ZERO, 0xCA)
          Opcode::ADDI(0x11, REG_ZERO, 0xBA)
          Opcode::LOG(0x10, 0x11, REG_ZERO, REG_ZERO)
          Opcode::RET(REG_ONE)
        */
        arrayify('0x504000ca504400ba3341100024040000'),
      scriptData: randomBytes(32),
      inputs: CoinInputs,
      witnesses: ['0x'],
    });

    const expectedReceipts: Receipt[] = [
      {
        type: ReceiptType.Log,
        id: ZeroBytes32,
        val0: bn(202),
        val1: bn(186),
        val2: bn(0),
        val3: bn(0),
        pc: bn(0x2870),
        is: bn(0x2868),
      },
      {
        type: ReceiptType.Return,
        id: ZeroBytes32,
        val: bn(1),
        pc: bn(0x2874),
        is: bn(0x2868),
      },
      {
        type: ReceiptType.ScriptResult,
        result: bn(0),
        gasUsed: bn(0x67),
      },
    ];

    expect(callResult.receipts).toStrictEqual(expectedReceipts);
  });

  // TODO: Add tests to provider sendTransaction
  // sendTransaction can't be tested without a valid signature
  // importing and testing it here can generate cycle dependency
  // as we test this in other modules like call contract its ok to
  // skip for now
  it.skip('can sendTransaction()', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);

    const response = await provider.sendTransaction({
      type: TransactionType.Script,
      gasPrice: 0,
      gasLimit: 1000000,
      script:
        /*
          Opcode::ADDI(0x10, REG_ZERO, 0xCA)
          Opcode::ADDI(0x11, REG_ZERO, 0xBA)
          Opcode::LOG(0x10, 0x11, REG_ZERO, REG_ZERO)
          Opcode::RET(REG_ONE)
        */
        arrayify('0x504000ca504400ba3341100024040000'),
      scriptData: randomBytes(32),
    });

    const result = await response.wait();

    expect(result.receipts).toEqual([
      {
        type: ReceiptType.Log,
        id: ZeroBytes32,
        val0: bn(202),
        val1: bn(186),
        val2: bn(0),
        val3: bn(0),
        pc: bn(0x2878),
        is: bn(0x2870),
      },
      {
        type: ReceiptType.Return,
        id: ZeroBytes32,
        val: bn(1),
        pc: bn(0x287c),
        is: bn(0x2870),
      },
      {
        type: ReceiptType.ScriptResult,
        result: bn(0),
        gasUsed: bn(0x2c),
      },
    ]);
  });

  it('can get all chain info', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);
    const { consensusParameters } = await provider.getChain();

    expect(consensusParameters.contractMaxSize).toBeDefined();
    expect(consensusParameters.maxInputs).toBeDefined();
    expect(consensusParameters.maxOutputs).toBeDefined();
    expect(consensusParameters.maxWitnesses).toBeDefined();
    expect(consensusParameters.maxGasPerTx).toBeDefined();
    expect(consensusParameters.maxScriptLength).toBeDefined();
    expect(consensusParameters.maxScriptDataLength).toBeDefined();
    expect(consensusParameters.maxStorageSlots).toBeDefined();
    expect(consensusParameters.maxPredicateLength).toBeDefined();
    expect(consensusParameters.maxPredicateDataLength).toBeDefined();
    expect(consensusParameters.maxGasPerPredicate).toBeDefined();
    expect(consensusParameters.gasPriceFactor).toBeDefined();
    expect(consensusParameters.gasPerByte).toBeDefined();
    expect(consensusParameters.maxMessageDataLength).toBeDefined();
  });

  it('can get node info including some consensus parameters properties', async () => {
    // #region provider-definition
    const provider = new Provider(FUEL_NETWORK_URL);
    const { minGasPrice, gasPerByte, gasPriceFactor, maxGasPerTx, nodeVersion } =
      await provider.getNodeInfo();
    // #endregion provider-definition

    expect(minGasPrice).toBeDefined();
    expect(gasPerByte).toBeDefined();
    expect(gasPriceFactor).toBeDefined();
    expect(maxGasPerTx).toBeDefined();
    expect(nodeVersion).toBeDefined();
  });

  it('can change the provider url of the current instance', () => {
    const providerUrl1 = FUEL_NETWORK_URL;
    const providerUrl2 = 'http://127.0.0.1:8080/graphql';
    const provider = new Provider(providerUrl1);
    const spyGraphQLClient = jest.spyOn(GraphQL, 'GraphQLClient');

    expect(provider.url).toBe(providerUrl1);
    provider.connect(providerUrl2);
    expect(provider.url).toBe(providerUrl2);
    expect(spyGraphQLClient).toBeCalledWith(providerUrl2, undefined);
  });

  it('can accept a custom fetch function', async () => {
    const providerUrl = FUEL_NETWORK_URL;

    const customFetch = async (
      url: string,
      options: {
        body: string;
        headers: { [key: string]: string };
        [key: string]: unknown;
      }
    ) => {
      const graphqlRequest = JSON.parse(options.body);
      const { operationName } = graphqlRequest;
      if (operationName === 'getVersion') {
        const responseText = JSON.stringify({
          data: { nodeInfo: { nodeVersion: '0.30.0' } },
        });
        const response = Promise.resolve(new Response(responseText, options));

        return response;
      }
      return fetch(url, options);
    };

    const provider = new Provider(providerUrl, { fetch: customFetch });
    expect(await provider.getVersion()).toEqual('0.30.0');
  });

  it('can force-produce blocks', async () => {
    // #region Provider-produce-blocks
    const provider = new Provider(FUEL_NETWORK_URL);

    const block = await provider.getBlock('latest');
    if (!block) {
      throw new Error('No latest block');
    }
    const { height: latestBlockNumberBeforeProduce } = block;

    const amountOfBlocksToProduce = 3;
    const latestBlockNumber = await provider.produceBlocks(amountOfBlocksToProduce);

    expect(latestBlockNumber.toHex()).toEqual(
      latestBlockNumberBeforeProduce.add(amountOfBlocksToProduce).toHex()
    );
    // #endregion Provider-produce-blocks
  });

  // TODO: Add back support for producing blocks with intervals by supporting the new
  // `block_production` config option for `fuel_core`.
  // See: https://github.com/FuelLabs/fuel-core/blob/def8878b986aedad8434f2d1abf059c8cbdbb8e2/crates/services/consensus_module/poa/src/config.rs#L20
  it.skip('can force-produce blocks with custom timestamps', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);

    const block = await provider.getBlock('latest');
    if (!block) {
      throw new Error('No latest block');
    }
    const { time: latestBlockTimestampBeforeProduce, height: latestBlockNumberBeforeProduce } =
      block;
    const latestBlockUnixTimestampBeforeProduce = fromTai64ToUnix(
      latestBlockTimestampBeforeProduce
    );

    const amountOfBlocksToProduce = 3;
    const blockTimeInterval = 100; // 100ms
    const startTime = new Date(latestBlockUnixTimestampBeforeProduce).getTime() + 1000; // 1s after the latest block

    const latestBlockNumber = await provider.produceBlocks(amountOfBlocksToProduce, startTime);

    // Verify that the latest block number is the expected one
    expect(latestBlockNumber.toString(10)).toEqual(
      latestBlockNumberBeforeProduce.add(amountOfBlocksToProduce).toString(10)
    );

    // Verify that the produced blocks have the expected timestamps and block numbers
    const producedBlocks = (
      await Promise.all(
        Array.from({ length: amountOfBlocksToProduce }, (_, i) =>
          provider.getBlock(latestBlockNumberBeforeProduce.add(i + 1).toNumber())
        )
      )
    ).map((producedBlock) => ({
      height: producedBlock?.height.toString(10),
      time: producedBlock?.time,
    }));
    const expectedBlocks = Array.from({ length: amountOfBlocksToProduce }, (_, i) => ({
      height: latestBlockNumberBeforeProduce.add(i + 1).toString(10),
      time: fromUnixToTai64(startTime + i * blockTimeInterval),
    }));
    expect(producedBlocks).toEqual(expectedBlocks);
  });

  it('can cacheUtxo [undefined]', () => {
    const provider = new Provider(FUEL_NETWORK_URL);

    expect(provider.cache).toEqual(undefined);
  });

  it('can cacheUtxo [numerical]', () => {
    const provider = new Provider(FUEL_NETWORK_URL, {
      cacheUtxo: 2500,
    });

    expect(provider.cache).toBeTruthy();
    expect(provider.cache?.ttl).toEqual(2_500);
  });

  it('can cacheUtxo [invalid numerical]', () => {
    expect(() => new Provider(FUEL_NETWORK_URL, { cacheUtxo: -500 })).toThrow(
      'Invalid TTL: -500. Use a value greater than zero.'
    );
  });

  it('can cacheUtxo [will not cache inputs if no cache]', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);
    const transactionRequest = new ScriptTransactionRequest({});

    const { error } = await safeExec(() => provider.sendTransaction(transactionRequest));

    expect(error).toBeTruthy();
    expect(provider.cache).toEqual(undefined);
  });

  it('can cacheUtxo [will not cache inputs cache enabled + no coins]', async () => {
    const provider = new Provider(FUEL_NETWORK_URL, {
      cacheUtxo: 1,
    });
    const MessageInput: MessageTransactionRequestInput = {
      type: InputType.Message,
      amount: 100,
      sender: BaseAssetId,
      recipient: BaseAssetId,
      witnessIndex: 1,
      nonce: BaseAssetId,
    };
    const transactionRequest = new ScriptTransactionRequest({
      inputs: [MessageInput],
    });

    const { error } = await safeExec(() => provider.sendTransaction(transactionRequest));

    expect(error).toBeTruthy();
    expect(provider.cache).toBeTruthy();
    expect(provider.cache?.getActiveData()).toStrictEqual([]);
  });

  it('can cacheUtxo [will cache inputs cache enabled + coins]', async () => {
    const provider = new Provider(FUEL_NETWORK_URL, {
      cacheUtxo: 10000,
    });
    const EXPECTED: BytesLike[] = [
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c500',
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c501',
      '0xda5d131c490db3868be9f8e228cf279bd98ef1de97129682777ed93fa088bc3f02',
    ];
    const MessageInput: MessageTransactionRequestInput = {
      type: InputType.Message,
      amount: 100,
      sender: BaseAssetId,
      recipient: BaseAssetId,
      witnessIndex: 1,
      nonce: BaseAssetId,
    };
    const CoinInputA: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[0],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputB: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: arrayify(EXPECTED[1]),
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputC: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[2],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const transactionRequest = new ScriptTransactionRequest({
      inputs: [MessageInput, CoinInputA, CoinInputB, CoinInputC],
    });

    const { error } = await safeExec(() => provider.sendTransaction(transactionRequest));

    expect(error).toBeTruthy();
    const EXCLUDED = provider.cache?.getActiveData() || [];
    expect(EXCLUDED.length).toEqual(3);
    expect(EXCLUDED.map((value) => hexlify(value))).toStrictEqual(EXPECTED);

    // clear cache
    EXCLUDED.forEach((value) => provider.cache?.del(value));
  });

  it('can cacheUtxo [will cache inputs and also use in exclude list]', async () => {
    const provider = new Provider(FUEL_NETWORK_URL, {
      cacheUtxo: 10000,
    });
    const EXPECTED: BytesLike[] = [
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c503',
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c504',
      '0xda5d131c490db3868be9f8e228cf279bd98ef1de97129682777ed93fa088bc3505',
    ];
    const MessageInput: MessageTransactionRequestInput = {
      type: InputType.Message,
      amount: 100,
      sender: BaseAssetId,
      recipient: BaseAssetId,
      witnessIndex: 1,
      nonce: BaseAssetId,
    };
    const CoinInputA: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[0],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputB: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: arrayify(EXPECTED[1]),
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputC: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[2],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const transactionRequest = new ScriptTransactionRequest({
      inputs: [MessageInput, CoinInputA, CoinInputB, CoinInputC],
    });

    const { error } = await safeExec(() => provider.sendTransaction(transactionRequest));

    expect(error).toBeTruthy();
    const EXCLUDED = provider.cache?.getActiveData() || [];
    expect(EXCLUDED.length).toEqual(3);
    expect(EXCLUDED.map((value) => hexlify(value))).toStrictEqual(EXPECTED);

    const owner = Address.fromRandom();
    const resourcesToSpendMock = jest.fn(() => Promise.resolve({ coinsToSpend: [] }));
    // @ts-expect-error mock
    provider.operations.getCoinsToSpend = resourcesToSpendMock;
    await provider.getResourcesToSpend(owner, []);

    expect(resourcesToSpendMock).toHaveBeenCalledWith({
      owner: owner.toB256(),
      queryPerAsset: [],
      excludedIds: {
        messages: [],
        utxos: EXPECTED,
      },
    });

    // clear cache
    EXCLUDED.forEach((value) => provider.cache?.del(value));
  });

  it('can cacheUtxo [will cache inputs cache enabled + coins]', async () => {
    const provider = new Provider(FUEL_NETWORK_URL, {
      cacheUtxo: 10000,
    });
    const EXPECTED: BytesLike[] = [
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c500',
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c501',
      '0xda5d131c490db3868be9f8e228cf279bd98ef1de97129682777ed93fa088bc3f02',
    ];
    const MessageInput: MessageTransactionRequestInput = {
      type: InputType.Message,
      amount: 100,
      sender: BaseAssetId,
      recipient: BaseAssetId,
      witnessIndex: 1,
      nonce: BaseAssetId,
    };
    const CoinInputA: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[0],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputB: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: arrayify(EXPECTED[1]),
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputC: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[2],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const transactionRequest = new ScriptTransactionRequest({
      inputs: [MessageInput, CoinInputA, CoinInputB, CoinInputC],
    });

    const { error } = await safeExec(() => provider.sendTransaction(transactionRequest));

    expect(error).toBeTruthy();
    const EXCLUDED = provider.cache?.getActiveData() || [];
    expect(EXCLUDED.length).toEqual(3);
    expect(EXCLUDED.map((value) => hexlify(value))).toStrictEqual(EXPECTED);

    // clear cache
    EXCLUDED.forEach((value) => provider.cache?.del(value));
  });

  it('can cacheUtxo [will cache inputs and also merge/de-dupe in exclude list]', async () => {
    const provider = new Provider(FUEL_NETWORK_URL, {
      cacheUtxo: 10000,
    });
    const EXPECTED: BytesLike[] = [
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c503',
      '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c504',
      '0xda5d131c490db3868be9f8e228cf279bd98ef1de97129682777ed93fa088bc3505',
    ];
    const MessageInput: MessageTransactionRequestInput = {
      type: InputType.Message,
      amount: 100,
      sender: BaseAssetId,
      recipient: BaseAssetId,
      witnessIndex: 1,
      nonce: BaseAssetId,
    };
    const CoinInputA: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[0],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputB: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: arrayify(EXPECTED[1]),
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const CoinInputC: CoinTransactionRequestInput = {
      type: InputType.Coin,
      id: EXPECTED[2],
      owner: BaseAssetId,
      assetId: BaseAssetId,
      txPointer: BaseAssetId,
      witnessIndex: 1,
      amount: 100,
    };
    const transactionRequest = new ScriptTransactionRequest({
      inputs: [MessageInput, CoinInputA, CoinInputB, CoinInputC],
    });

    const { error } = await safeExec(() => provider.sendTransaction(transactionRequest));

    expect(error).toBeTruthy();
    const EXCLUDED = provider.cache?.getActiveData() || [];
    expect(EXCLUDED.length).toEqual(3);
    expect(EXCLUDED.map((value) => hexlify(value))).toStrictEqual(EXPECTED);

    const owner = Address.fromRandom();
    const resourcesToSpendMock = jest.fn(() => Promise.resolve({ coinsToSpend: [] }));
    // @ts-expect-error mock
    provider.operations.getCoinsToSpend = resourcesToSpendMock;
    await provider.getResourcesToSpend(owner, [], {
      utxos: [
        '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c503',
        '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c507',
        '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c508',
      ],
    });

    expect(resourcesToSpendMock).toHaveBeenCalledWith({
      owner: owner.toB256(),
      queryPerAsset: [],
      excludedIds: {
        messages: [],
        utxos: [
          '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c503',
          '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c507',
          '0xbc90ada45d89ec6648f8304eaf8fa2b03384d3c0efabc192b849658f4689b9c508',
          EXPECTED[1],
          EXPECTED[2],
        ],
      },
    });

    // clear cache
    EXCLUDED.forEach((value) => provider.cache?.del(value));
  });

  it('can getBlocks', async () => {
    const provider = new Provider(FUEL_NETWORK_URL);
    // Force-producing some blocks to make sure that 10 blocks exist
    await provider.produceBlocks(10);
    // #region Provider-get-blocks
    const blocks = await provider.getBlocks({
      last: 10,
    });
    // #endregion Provider-get-blocks
    expect(blocks.length).toBe(10);
    blocks.forEach((block) => {
      expect(block).toEqual(
        expect.objectContaining({
          id: expect.any(String),
          height: expect.any(BN),
          time: expect.any(String),
          transactionIds: expect.any(Array<string>),
        })
      );
    });
  });

  it('can getMessageProof with all data', async () => {
    // Create a mock provider to return the message proof
    // It test mainly types and converstions
    const provider = new Provider(FUEL_NETWORK_URL, {
      fetch: async (url, options) => {
        const messageProof = JSON.stringify(messageProofResponse);
        return Promise.resolve(new Response(messageProof, options));
      },
    });
    const messageProof = await provider.getMessageProof(
      '0x79c54219a5c910979e5e4c2728df163fa654a1fe03843e6af59daa2c3fcd42ea',
      '0xb33895e6fdf23b5a62c92a1d45c71a11579027f9e5c4dda73c26cf140bcd6895',
      '0xe4dfe8fc1b5de2c669efbcc5e4c0a61db175d1b2f03e3cd46ed4396e76695c5b'
    );
    expect(messageProof).toMatchSnapshot();
  });
});
