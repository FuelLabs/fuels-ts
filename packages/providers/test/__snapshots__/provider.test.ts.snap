// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Provider can getMessageProof with all data 1`] = `
{
  "amount": "0xa",
  "blockProof": {
    "proofIndex": "0x1b0",
    "proofSet": [
      "0x3480098c927d3a8a86745ba6c91419f04901baf5d25bbd0e6ba56fcb3b3e35bd",
      "0xa8795e5d15b88c2383f5673a4658edae93dcdb7986ccfacbe916e656ec11ce70",
      "0x97ea3aee8cb9ccc6593c21879979eeab0f748066b9e22e2294cb973ee5d95e3c",
      "0x50a33f4fe58e35955df499092515fe2f77d7bfaed35f80194b338d7091c363ab",
      "0xa07fc684d351ab1970efe9ffd04eb67b6544dd4443169ecf6dd0bfc546216211",
    ],
  },
  "commitBlockHeader": {
    "applicationHash": "0x441dece8e41d111f3a875ed89d716e48abf450ad385b111cf7c55e06e0b57d6b",
    "daHeight": "0x0",
    "height": "0x1b1",
    "id": "0xe4dfe8fc1b5de2c669efbcc5e4c0a61db175d1b2f03e3cd46ed4396e76695c5b",
    "messageReceiptCount": "0x0",
    "messageReceiptRoot": "0xe3b0c44298fc1c149afbf4c8996fb92427ae41e4649b934ca495991b7852b855",
    "prevRoot": "0xcc024475c2ccd175082bc233f5ebb3736e47a648a666d82567b001ab46cf0a83",
    "time": "4611686020115956241",
    "transactionsCount": "0x2",
    "transactionsRoot": "0x471d68713ae2fef6bb3eede37a976b38390f2ab8916e463b22942e7130456a8a",
  },
  "data": "0x",
  "messageBlockHeader": {
    "applicationHash": "0x76b6838a38a3049fa1dd42f923a0f287594a390493003af0f1c3da94f3d2b8c7",
    "daHeight": "0x0",
    "height": "0x1b0",
    "id": "0x864b55089878bf7009d2ff64cbeeeeb75fcd73768785dcb75d54180e7fbaab7b",
    "messageReceiptCount": "0x1",
    "messageReceiptRoot": "0x34fdfbc73458998a106b02f13bba51bc0fd36513602e312ca1d1893f2075eca5",
    "prevRoot": "0x26e49e4af2253e0bca68d2a8b38fa7406a546ebc476b6ca51569cb325ad55ea6",
    "time": "4611686020115956241",
    "transactionsCount": "0x2",
    "transactionsRoot": "0xbf760e126159b72f59bde2d59a972620f826e5a5134deb2a775a21c37644e1a1",
  },
  "messageProof": {
    "proofIndex": "0x0",
    "proofSet": [
      "0x34fdfbc73458998a106b02f13bba51bc0fd36513602e312ca1d1893f2075eca5",
    ],
  },
  "nonce": "0x3e87e0f44613cabecd1aad381ad41a433afb12ec5c54c172de3db25b1b4d1b53",
  "recipient": "fuel1qqqqqqqqqqqqqqqqqqqy0wnpamywtejjglt30lerdagyeuas5f3s0esvs7",
  "sender": "fuel108z5yxd9eygf08j7fsnj3hck87n9fg07qwzru6h4nk4zc07dgt4q5t8nka",
}
`;
